File: app/csharp/tests/Ch03/ImmutableOperationsTests.cs
using Ch03;
using LanguageExt;
using static LanguageExt.Prelude;
using Xunit;

namespace GrokkingFP.CSharp.Tests.Ch03;

/// <summary>
/// 第3章: イミュータブルなデータ操作のテスト
/// </summary>
public class ImmutableOperationsTests
{
    // ============================================
    // 基本的なリスト操作のテスト
    // ============================================

    [Fact]
    public void FirstN_ReturnsFirstNElements()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(1, 2, 3), ImmutableOperations.FirstN(seq, 3));
    }

    [Fact]
    public void FirstN_ReturnsAllWhenNIsLarger()
    {
        var seq = Seq(1, 2);
        Assert.Equal(Seq(1, 2), ImmutableOperations.FirstN(seq, 5));
    }

    [Fact]
    public void LastN_ReturnsLastNElements()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(3, 4, 5), ImmutableOperations.LastN(seq, 3));
    }

    [Fact]
    public void LastN_ReturnsAllWhenNIsLarger()
    {
        var seq = Seq(1, 2);
        Assert.Equal(Seq(1, 2), ImmutableOperations.LastN(seq, 5));
    }

    [Fact]
    public void FirstTwo_ReturnsFirstTwoElements()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(1, 2), ImmutableOperations.FirstTwo(seq));
    }

    [Fact]
    public void LastTwo_ReturnsLastTwoElements()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(4, 5), ImmutableOperations.LastTwo(seq));
    }

    [Fact]
    public void Slice_ReturnsElementsInRange()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(2, 3, 4), ImmutableOperations.Slice(seq, 1, 4));
    }

    // ============================================
    // 要素の追加のテスト
    // ============================================

    [Fact]
    public void Appended_AddsElementAtEnd()
    {
        var seq = Seq(1, 2, 3);
        Assert.Equal(Seq(1, 2, 3, 4), ImmutableOperations.Appended(seq, 4));
    }

    [Fact]
    public void AppendedAll_AddsAllElementsAtEnd()
    {
        var seq = Seq(1, 2);
        var elements = Seq(3, 4, 5);
        Assert.Equal(Seq(1, 2, 3, 4, 5), ImmutableOperations.AppendedAll(seq, elements));
    }

    // ============================================
    // リスト変換パターンのテスト
    // ============================================

    [Fact]
    public void MoveFirstTwoToEnd_MovesElements()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(3, 4, 5, 1, 2), ImmutableOperations.MoveFirstTwoToEnd(seq));
    }

    [Fact]
    public void InsertBeforeLast_InsertsElement()
    {
        var seq = Seq(1, 2, 3);
        Assert.Equal(Seq(1, 2, 99, 3), ImmutableOperations.InsertBeforeLast(seq, 99));
    }

    [Fact]
    public void InsertAtMiddle_InsertsAtMiddle()
    {
        var seq = Seq(1, 2, 3, 4);
        Assert.Equal(Seq(1, 2, 99, 3, 4), ImmutableOperations.InsertAtMiddle(seq, 99));
    }

    [Fact]
    public void InsertAtMiddle_InsertsAtMiddleOddLength()
    {
        var seq = Seq(1, 2, 3, 4, 5);
        Assert.Equal(Seq(1, 2, 99, 3, 4, 5), ImmutableOperations.InsertAtMiddle(seq, 99));
    }

    // ============================================
    // イミュータビリティの確認
    // ============================================

    [Fact]
    public void Operations_DoNotMutateOriginal()
    {
        var original = Seq(1, 2, 3);
        var _ = ImmutableOperations.Appended(original, 4);
        Assert.Equal(3, original.Count);
    }
}

/// <summary>
/// 旅程の再計画のテスト
/// </summary>
public class ItineraryTests
{
    [Fact]
    public void Replan_InsertsBeforeSpecifiedCity()
    {
        var plan = Seq("Tokyo", "Osaka", "Kyoto");
        var result = Itinerary.Replan(plan, "Nagoya", "Osaka");
        Assert.Equal(Seq("Tokyo", "Nagoya", "Osaka", "Kyoto"), result);
    }

    [Fact]
    public void Replan_AddsToEndWhenCityNotFound()
    {
        var plan = Seq("Tokyo", "Osaka");
        var result = Itinerary.Replan(plan, "Kyoto", "NotFound");
        Assert.Equal(Seq("Tokyo", "Osaka", "Kyoto"), result);
    }

    [Fact]
    public void Replan_InsertsAtBeginningWhenBeforeFirstCity()
    {
        var plan = Seq("Tokyo", "Osaka");
        var result = Itinerary.Replan(plan, "Nagoya", "Tokyo");
        Assert.Equal(Seq("Nagoya", "Tokyo", "Osaka"), result);
    }
}

/// <summary>
/// 文字列操作のテスト
/// </summary>
public class StringOperationsTests
{
    [Fact]
    public void Abbreviate_ConvertsToInitialDotLastName()
    {
        Assert.Equal("A. Church", StringOperations.Abbreviate("Alonzo Church"));
        Assert.Equal("J. McCarthy", StringOperations.Abbreviate("John McCarthy"));
    }

    [Fact]
    public void Abbreviate_ReturnsOriginalWhenNoSpace()
    {
        Assert.Equal("Alonzo", StringOperations.Abbreviate("Alonzo"));
    }

    [Fact]
    public void FirstNChars_ReturnsFirstNCharacters()
    {
        Assert.Equal("Hel", StringOperations.FirstNChars("Hello", 3));
        Assert.Equal("Hi", StringOperations.FirstNChars("Hi", 5));
    }

    [Fact]
    public void LastNChars_ReturnsLastNCharacters()
    {
        Assert.Equal("llo", StringOperations.LastNChars("Hello", 3));
        Assert.Equal("Hi", StringOperations.LastNChars("Hi", 5));
    }
}

File: app/fsharp/tests/Ch03/ImmutableOperationsTests.fs
module Ch03.ImmutableOperationsTests

open Xunit
open Ch03.ImmutableOperations
open Ch03.Itinerary
open Ch03.StringOperations

// ============================================
// 基本的なリスト操作のテスト
// ============================================

[<Fact>]
let ``firstTwo は最初の2要素を返す`` () =
    Assert.Equal<string list>(["a"; "b"], firstTwo ["a"; "b"; "c"])
    Assert.Equal<string list>(["a"; "b"], firstTwo ["a"; "b"])
    Assert.Equal<string list>(["a"], firstTwo ["a"])
    Assert.Equal<string list>([], firstTwo ([] : string list))

[<Fact>]
let ``lastTwo は最後の2要素を返す`` () =
    Assert.Equal<string list>(["b"; "c"], lastTwo ["a"; "b"; "c"])
    Assert.Equal<string list>(["a"; "b"], lastTwo ["a"; "b"])
    Assert.Equal<string list>(["a"], lastTwo ["a"])
    Assert.Equal<string list>([], lastTwo ([] : string list))

[<Fact>]
let ``slice はリストの一部を切り出す`` () =
    Assert.Equal<string list>(["b"; "c"], slice 1 3 ["a"; "b"; "c"; "d"])
    Assert.Equal<string list>(["a"; "b"], slice 0 2 ["a"; "b"; "c"])
    Assert.Equal<string list>([], slice 0 0 ["a"; "b"; "c"])

// ============================================
// 要素の追加のテスト
// ============================================

[<Fact>]
let ``appended はリストの末尾に要素を追加する`` () =
    let appleBook = ["Apple"; "Book"]
    let appleBookMango = appended "Mango" appleBook

    // 元のリストは変わらない（イミュータブル）
    Assert.Equal(2, List.length appleBook)
    Assert.Equal(3, List.length appleBookMango)
    Assert.Equal<string list>(["Apple"; "Book"; "Mango"], appleBookMango)

[<Fact>]
let ``appendedAll はリストの末尾に複数の要素を追加する`` () =
    let ab = ["a"; "b"]
    let cd = ["c"; "d"]
    let abcd = appendedAll cd ab

    Assert.Equal<string list>(["a"; "b"; "c"; "d"], abcd)

// ============================================
// リスト変換パターンのテスト
// ============================================

[<Fact>]
let ``moveFirstTwoToEnd は最初の2要素を末尾に移動する`` () =
    Assert.Equal<string list>(["c"; "a"; "b"], moveFirstTwoToEnd ["a"; "b"; "c"])
    Assert.Equal<string list>(["c"; "d"; "a"; "b"], moveFirstTwoToEnd ["a"; "b"; "c"; "d"])

[<Fact>]
let ``insertBeforeLast は最後の要素の前に挿入する`` () =
    Assert.Equal<string list>(["a"; "c"; "b"], insertBeforeLast "c" ["a"; "b"])
    Assert.Equal<string list>(["a"; "b"; "X"; "c"], insertBeforeLast "X" ["a"; "b"; "c"])

[<Fact>]
let ``insertAtMiddle は中央に要素を挿入する`` () =
    Assert.Equal<string list>(["a"; "b"; "X"; "c"; "d"], insertAtMiddle "X" ["a"; "b"; "c"; "d"])
    Assert.Equal<string list>(["a"; "X"; "b"], insertAtMiddle "X" ["a"; "b"])
    Assert.Equal<string list>(["X"; "a"], insertAtMiddle "X" ["a"])

// ============================================
// 旅程の再計画のテスト
// ============================================

[<Fact>]
let ``replan は指定した都市の前に新しい都市を挿入する`` () =
    let planA = ["Paris"; "Berlin"; "Kraków"]
    let planB = replan planA "Vienna" "Kraków"

    // Plan B には Vienna が追加されている
    Assert.Equal<string list>(["Paris"; "Berlin"; "Vienna"; "Kraków"], planB)

    // Plan A は変更されていない（イミュータブル）
    Assert.Equal<string list>(["Paris"; "Berlin"; "Kraków"], planA)

[<Fact>]
let ``replan は都市が見つからない場合末尾に追加する`` () =
    let plan = ["Paris"; "Berlin"]
    let result = replan plan "Vienna" "NotFound"

    Assert.Equal<string list>(["Paris"; "Berlin"; "Vienna"], result)

// ============================================
// 文字列操作のテスト
// ============================================

[<Fact>]
let ``abbreviate は名前を省略形に変換する`` () =
    Assert.Equal("A. Church", abbreviate "Alonzo Church")
    Assert.Equal("A. Turing", abbreviate "Alan Turing")

[<Fact>]
let ``abbreviate は既に省略形の場合はそのまま返す`` () =
    Assert.Equal("A. Church", abbreviate "A. Church")

[<Fact>]
let ``abbreviate はスペースがない場合はそのまま返す`` () =
    Assert.Equal("Alonzo", abbreviate "Alonzo")

[<Fact>]
let ``firstNChars は最初のn文字を返す`` () =
    Assert.Equal("Hel", firstNChars 3 "Hello")
    Assert.Equal("Hello", firstNChars 10 "Hello")
    Assert.Equal("", firstNChars 0 "Hello")

[<Fact>]
let ``lastNChars は最後のn文字を返す`` () =
    Assert.Equal("llo", lastNChars 3 "Hello")
    Assert.Equal("Hello", lastNChars 10 "Hello")
    Assert.Equal("", lastNChars 0 "Hello")
